### ISSUE TITLE
Support for non integer offset values

### ISSUE THREAD

**Issue by manquer** [yiisoft/yii2#8969]
Some databases support functions/ sub query for determining offset values like 
for ex in postgres this is allowed
`SELECT myid FROM mytable OFFSET floor(random()*N) ;`
or
`SELECT myid FROM mytable OFFSET (SELECT COUNT(myid)-100 FROM MYtable) ;`

However yii\db\QueryTrait supports only integer offsets
Support for function based/ subquery based offset values, alternatively non integer values should not be allowed in the function argument

**Comment by samdark** [comment]
I guess it could be done with Expression.

**Comment by klimov-paul** [comment]
There is a little problem: in order to support `yi\\db\Expression` with params an additional argument should be added to the method `QueryBuilder::buildLimit()`, which actually breaks a BC.

**Comment by SamMousa** [comment]
@klimov-paul Is that really needed?
For `orderBy` it is solved by adding the params after building the sql snippet:
````
$sql = $this->buildOrderByAndLimit($sql, $query->orderBy, $query->limit, $query->offset);

        if (!empty($query->orderBy)) {
            foreach ($query->orderBy as $expression) {
                if ($expression instanceof Expression) {
                    $params = array_merge($params, $expression->params);
                }
            }
        }
````

It seems like the same could be done for limit and offset?

**Comment by samdark** [comment]
Won't be implemented in 2.0.
